src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -150,8 +152,29 @@ private void AddRange(IEnumerable<KeyValuePair<TKey, TValue>> collection)
                 return;
             }
 
-            // Fallback path for IEnumerable that isn't a non-subclassed Dictionary<TKey,TValue>.
-            foreach (KeyValuePair<TKey, TValue> pair in collection)
+            // We similarly special-case KVP<>[] and List<KVP<>>, as they're commonly used to seed dictionaries, and
```

Comment:
Would it make sense to just specialize for `IList<KeyValuePair<TKey, TValue>>`? It would incur $O(n)$ virtual calls but maybe it's not the dominant factor in this particular case? It would also extend the optimization to other collections as well.